src/nnet3/nnet-simple-component.cc

```diff
@@ -119,16 +124,28 @@ void DropoutComponent::Propagate(const ComponentPrecomputedIndexes *indexes,
 
   BaseFloat dropout = dropout_proportion_;
   KALDI_ASSERT(dropout >= 0.0 && dropout <= 1.0);
-
-  // This const_cast is only safe assuming you don't attempt
-  // to use multi-threaded code with the GPU.
-  const_cast<CuRand<BaseFloat>&>(random_generator_).RandUniform(out);
-
-  out->Add(-dropout); // now, a proportion "dropout" will be <0.0
-  out->ApplyHeaviside(); // apply the function (x>0?1:0).  Now, a proportion "dropout" will
-                         // be zero and (1 - dropout) will be 1.0.
-
-  out->MulElements(in);
+  if (!dropout_per_frame_) {
+    // This const_cast is only safe assuming you don't attempt
+    // to use multi-threaded code with the GPU.
+    const_cast<CuRand<BaseFloat>&>(random_generator_).RandUniform(out);
+
+    out->Add(-dropout);  // now, a proportion "dropout" will be <0.0
+    out->ApplyHeaviside();  // apply the function (x>0?1:0).  Now, a proportion
+                           // "dropout" will be zero and (1 - dropout) will be 1.0.
+    out->MulElements(in);
+  } else {
+    // randomize the dropout matrix by row,
+    // i.e. [[1,1,1,1],[0,0,0,0],[0,0,0,0],[1,1,1,1],[0,0,0,0]]
+    CuMatrix<BaseFloat> tmp(1, out->NumRows(), kUndefined);
```

Comment:
You should now be able to change this code to:
(a) use a vector instead of a matrix for 'tmp', and
(b) use CopyVecToCols() instead of AddVecToCols(), which avoids the need to call SetZero().

In about 20 minutes I'll push @kangshiyin's changes regarding implementing CopyVecToCols() to the master branch; and @vesis84 has already implemented RandUniform for vectors